### ISSUE TITLE
Enzyme cast shadow argument error

### ISSUE THREAD

**Issue by jpmedinagl** [EnzymeAD/rust#187]
Enzyme cast shadow argument error
```
[compiler/rustc_codegen_llvm/src/back/write.rs:1234:9] "Running Module Optimization after differentiation" = "Running Module Optimization after differentiation"
error: <unknown>:0:0: in function _ZN3opt5dcart17h15b297d77510e859E void (ptr, i64, ptr, i64, ptr, ptr, i64, i32, ptr, i64, i32, ptr, i64): Enzyme: Cannot cast __enzyme_autodiff shadow argument 3, found i64 %1, type i64 - to arg 0 ptr
```

Here is the function header and autodiff macros
```
#![feature(autodiff)]

use std::autodiff::autodiff;

#[no_mangle]
#[autodiff(dcart, Reverse, Duplicated, Const, Const, Const, Const, Const, Duplicated)]
fn cart(
    out: &mut [f64],
    shls: [i32; 4],
    atm: &mut [i32],
    natm: i32,
    bas: &mut [i32],
    nbas: i32,
    env: &mut [f64],
)
```

**Comment by ZuseZ4** [comment]
```rs
#![feature(autodiff)]

use std::autodiff::autodiff;

#[no_mangle]
#[autodiff(dcart, Reverse, Duplicated, Const, Const, Const)]
fn cart(
    out: &mut [f64],
    shls: [i32; 4],
    atm: &mut [i32],
    natm: i32,
){}

fn main() {
    let shls = [0, 1, 2, 3];
    let mut atm = [0, 1, 2, 3];
    let mut out = [0.0, 0.0, 0.0, 0.0];
    let mut dout = [0.0, 0.0, 0.0, 0.0];
    let natm = 4;
    dcart(&mut out, &mut dout, shls, &mut atm, natm);
}
```

**Comment by ZuseZ4** [comment]
closed by #a943e179a2d99b69248b077a30725914f4919fbc in the opt2 branch.
